File: python/mrpt.py
# encoding: utf-8
#
# Author: Teemu Henrikki Pitkänen <teemu.pitkanen@helsinki.fi>
# University of Helsinki / Helsinki Institute for Information Technology 2016
#

import numpy as np
from scipy.spatial.distance import cdist
from Queue import PriorityQueue, Empty
from collections import deque
from math import log, ceil, floor

# The following imports used only for saving/loading trees
import os
import cPickle
from hashlib import sha1


class MRPTIndex(object):
    """
    The MRPT index is a data structure that allows us to answer approximate nearest neighbor queries quickly. The index
    is basically just a collection of random projection trees. This class implements the basic MRPT index as well as a
    number of small tricks to improve performance.
    """
    def __init__(self, data, n0, n_trees, degree=2, use_saved=False):
        """
        The initializer builds the MRPT index. It has to be called before ANN-queries can be made.
        :param data: The data fro which the index is built. At the moment this has to be done at once, no objects can be
        added later.
        :param n0: The maximum leaf size used in all of the trees of the index.
        :param n_trees: The number of trees used in the index.
        :param degree: Defines the degree of inner random projection tree nodes (the number of children). Use the
        default value unless you really know what you are doing.
        :param use_saved: Saves and loads trees for reuse instead of building new ones every time an index is created.
        :return: The index object.
        """
        self.data = data
        if use_saved:
            self.trees = []
            # The directory name is a hash of the data set contents to avoid using trees built for wrong data
            save_path = 'saved_trees/'+sha1(data.view(np.uint8)).hexdigest()[:8]+'/'+str(n0)+'/'+str(degree)
            self.trees = MRPTIndex.load_trees(save_path, n_trees)
            for t in range(len(self.trees), n_trees):
                t = RPTree(data, n0, degree=degree)
                self.trees.append(t)
                MRPTIndex.save_trees([t], save_path)
        else:
            self.trees = [RPTree(data, n0) for t in range(n_trees)]

    def ann(self, obj, k, extra_branches=0, n_elected=0):
        """
        The mrpt approximate nearest neighbor query. By default the function implements a basic query - the query object
        is routed to one leaf in each tree, and the nearest neighbors are searched in the union of these leaves. The
        extra parameters allow some tricks to improve performance. The parameter extra_brances allows the query to
        explore extra branches where the projection value was close to the split. The n_elected parameter causes the
        query to use the voting trick - not all objects in the union of leaves are searched, but the appearance counts
        as a vote and only the most voted objects get 'elected' to be searched. Thus an object must be returned by
        several trees to be included in the linear search.
        :param obj: The object whose neighbors are being searched
        :param k: The number of neighbors being searched
        :param extra_branches: The number of extra branches allowed in the index
        :param n_elected: The number of elected objects in the voting trick
        :return: The approximate neighbors. In extreme situations not strictly k, but slightly less (eg. 1 tree case)
        """
        priority_queue = PriorityQueue()
        all_projections = []
        votes = np.zeros(len(self.data))

        # First traverse each tree from root to leaf
        for tree_id in range(len(self.trees)):
            indexes, gaps, projections = self.trees[tree_id].full_tree_traversal(obj)
            votes[indexes] += 1
            all_projections.append(projections)
            for (gap_width, node, level) in gaps:
                priority_queue.put((gap_width, node, level, tree_id))

        # Optional branching trick: traverse down from #extra_branches nodes with the smallest d(projection, split)
        for i in range(extra_branches):
            try:
                gap_width, node, level, tree = priority_queue.get(block=False)
                indexes, gaps = RPTree.partial_tree_traversal(node, all_projections[tree][level:], level)
                votes[indexes] += 1
                for gap in gaps:
                    priority_queue.put((gap[0], gap[1], gap[2], tree))
            except Empty:
                print 'More branches than leaves. Will skip the extras.'

        # Decide which nodes to include in the brute force search
        if n_elected > 0:   # Optional voting trick
            elected = np.argsort(votes)[::-1][:n_elected]
        else:  # Basic mrpt
            elected = np.nonzero(votes)[0]

        # Find the nearest neighbors in the subset of objects
        return elected[np.argsort(cdist([obj], self.data[elected])[0])[:k]]

    @staticmethod
    def save_trees(trees, path):
        """
        A function for saving the trees to disk.
        :param trees: The trees to be saved
        :param path: The path where the trees are saved
        """
        if not os.path.exists(path):
            os.makedirs(path)
        ordinal = 0
        for tree in trees:
            while os.path.isfile(path + '/t' + str(ordinal) + '.idx'):
                ordinal += 1
            filename = path + '/t' + str(ordinal) + '.idx'
            with open(filename, 'w') as f:
                cPickle.dump(tree, f)

    @staticmethod
    def load_trees(path, n_trees):
        """
        A function for reading saved trees from the disk.
        :param path: The path from where the trees are loaded
        :param n_trees: The number of trees the user wants loaded
        :return: A list containing the trees. Empty if no such directory exists. If not enough trees exist the function
        will return as many as it finds.
        """
        trees = []
        if os.path.exists(path):
            files = os.listdir(path)
            for i in range(min(n_trees, len(files))):
                with open(path+'/'+files[i], 'r') as f:
                    trees.append(cPickle.load(f))
        return trees


class RPTree(object):
    """
    A Random Projection Tree is a spatial index data structure used in specific data mining and indexing tasks. Our
    implementation has been built with approximate nearest neighbor search (ANN) problems in mind. A tree is built by
    dividing the data space by random hyperplanes into small cells. In ANN-problems we usually achieve substantial
    improvements by quickly choosing just a subset of the data located in a single cell where the actual brute-force NN
    search is then performed, instead of using the whole data set.
    """

    def __init__(self, data, n0, degree=2):
        """
        Sets the main attributes and calls the build_tree routine which actually creates the nodes and builds the tree.
        :param data: The data for which the index is built
        :param n0: The maximum leaf size of the tree
        :param degree: The maximum number of children that each internal node of the tree has
        """
        self.seed = np.random.randint(0, int(1e9))
        self.degree = degree
        self.tree_height = int(ceil(log(len(data)/float(n0), degree)))
        self.root = _Node()
        self._build_tree(data, n0)

    def _build_tree(self, data, n0):
        """
        An iterative method for building the random projection tree structure. The tree is built level-by-level by
        using a queue to handle the order in which the nodes are processed. The method is significantly faster to
        recursive building. This implementation uses same random vectors in each tree branch for running
        time efficiency. Note that although all projections are computed in a single matrix multiplication, the
        projection vector is different on each level.
        """
        # Restore rng settings for reproducibility and compute projections to random basis
        np.random.seed(self.seed)
        all_projections = np.dot(data, np.random.normal(size=(data.shape[1], self.tree_height)))

        # Main while loop that builds the tree one level at a time
        queue = deque([(self.root, np.arange(len(data)))])
        tracker = _FullTreeLevelTracker(self.degree)
        while len(queue) > 0:
            # Pop next node to be processed
            node, indexes = queue.popleft()

            # Divide the indexes into equal sized chunks (one for each child) and compute the split boundaries
            indexes_divided, node.splits = self._chunkify(indexes, all_projections[indexes, tracker.level], n0)

            # Set references to children, add child nodes to queue if further splits are required (node size > n0)
            for node_indexes in indexes_divided:
                if len(node_indexes) > n0:
                    child = _Node()
                    node.children.append(child)
                    queue.append((child, node_indexes))
                else:
                    node.children.append(node_indexes)

            tracker.object_added()  # Corresponds to adding _node_, not its children, thus called only once

    def _chunkify(self, indexes, projections, n0):
        """
        Divides the list 'indexes' into 'self.degree' equal-sized chunks. If the split is not even, the extra elements
        are added to the list entries on the left. The function aims at building as full leaf nodes as possible, so in
        case each child is not of size at least n0, the method may result in less than 'self.degree' chunks. Think of
        the case n0=10, self.degree=5, len(indexes)=12. Instead of splitting into 5 chunks of sizes 2-3, the function
        returns two chunks with sizes 6 and 5. Thus the leaf sizes are generally closer to n0 and the running time and
        performance stay more predictable.
        :param indexes: The list of indexes to be chunkified
        :param projections: The projections corresponding to the indexes
        :param n0: The maximum leaf size
        :return: The chunks as a list of lists, the projection values that separate these chunks
        """
        indexes = indexes[np.argsort(projections)]
        projections = np.sort(projections)
        n = len(indexes)

        n_chunks = min(int(ceil(n/float(n0))), self.degree)
        min_chunk_size = int(floor(n/n_chunks))
        chunk_sizes = np.repeat([min_chunk_size], n_chunks)
        chunk_sizes[range(n - min_chunk_size*n_chunks)] += 1
        chunk_bounds = np.cumsum(np.concatenate(([0], chunk_sizes)))

        return ([indexes[chunk_bounds[i]:chunk_bounds[i+1]] for i in range(n_chunks)],
                [(projections[i-1] + projections[i])/2 for i in chunk_bounds[1:-1]])

    def full_tree_traversal(self, obj):
        """
        The function places the query object 'obj' to a leaf. The function re-creates the same random vectors that were
        used in tree construction, computes the projections of the query vector and using the split information stored
        in the nodes places the query vector into a single leaf.
        :param obj: The query object, has to be given as a row vector
        :return: The indexes of the leaf, gap information and projection values from the path to leaf (for the extra
        branches trick)
        """
        # Restore rng settings, compute projections to random basis
        np.random.seed(self.seed)
        projections = np.dot(obj, np.random.normal(size=(len(obj), self.tree_height)))

        # Move down the tree according to the projections and split values stored in the tree
        indexes, gaps = self.partial_tree_traversal(self.root, projections, 0)
        return indexes, gaps, projections

    @staticmethod
    def partial_tree_traversal(node, projections, tree_level):
        """
        Moves down to a leaf starting from the specified node.
        """
        gaps = []
        for projection in projections:
            # Leaves as
            if not hasattr(node, 'splits'):
                break

            # Find the child where the query object belongs.
            child_index = len(node.splits)
            for i in range(len(node.splits)):
                if projection < node.splits[i]:
                    child_index = i
                    break

            # Store the distances to splits for the priority queue trick.
            for i in range(len(node.splits)):
                gap = abs(projection - node.splits[i])
                if i < child_index:
                    gaps.append((gap, node.children[i], tree_level + 1))
                elif i >= child_index:
                    gaps.append((gap, node.children[i+1], tree_level + 1))

            # Move down the tree for next iteration round
            node = node.children[child_index]
            tree_level += 1

        return node, gaps


class _FullTreeLevelTracker(object):
    """
    A _FullTreeLevelTracker object keeps track on the height of a tree with a fixed degree in internal nodes. Every time
    an object is added to the tree the object_added -method need to be called. The height can be read from the 'level'
    attribute.
    """
    def __init__(self, degree=2):
        self.level = 0
        self.level_capacity = 1
        self.level_occupancy = 0
        self.degree = degree

    def object_added(self):
        """
        Must be called every time a node is added to the tree, updates the attributes accordingly.
        """
        self.level_occupancy += 1
        if self.level_occupancy == self.level_capacity:
            self.level_occupancy = 0
            self.level_capacity *= self.degree
            self.level += 1


class _Node(object):
    """
    The class describes the structure of a single internal node. Only the split values used at this node and the links
    to the child nodes need to be stored. (The random vector can be generated without explicitly storing it.). Notice
    also that leaf nodes are not _Node-objects but just simple lists of data object indices.
    """
    def __init__(self):
        self.children = []
        self.splits = []

File: cpp/Mrpt.cpp
/********************************************************
 * Ville Hyvönen & Teemu Pitkänen                       *
 * HIIT / University of Helsinki                        *
 * ville.o.hyvonen<at>helsinki.fi                       *
 * teemu.pitkanen<at>cs.helsinki.fi                     *
 * 2016                                                 *
 ********************************************************/

#include "armadillo"
#include <cstdlib>
#include <queue>
#include "Mrpt.h"

using namespace arma;

/**
 * The constructor of the index. The inputs are the data for which the index 
 * will be built and additional parameters that affect the accuracy of the NN
 * approximation. Concisely, larger n_trees_ or smaller depth values improve 
 * accuracy but slow down the queries. A general rule for the right balance is 
 * not known. The constructor does not actually build the trees, but that is 
 * done by a separate function 'grow' that has to be called before queries can 
 * be made. 
 * @param X_ - The data to be indexed. Samples as columns, features as rows.
 * @param n_trees_ - The number of trees to be used in the index.
 * @param depth_ - The depth of the trees
 * @param id_ - A name used for filenames when saving.
 */
Mrpt::Mrpt(const fmat& X_, int n_trees_, int depth_, std::string id_) : X(X_), n_trees(n_trees_), depth(depth_), id(id_){
    n_samples = X.n_cols; 
    dim = X.n_rows;
    n_pool = n_trees * depth;
    n_array = pow(2, depth + 1);
    split_points = fmat();
    random_matrix = fmat();
}

Mrpt::Mrpt(const std::string filename, int n_trees_, int depth_, std::string id_) : n_trees(n_trees_), depth(depth_), id(id_){
    X = fmat();
    X.load(filename);
    n_samples = X.n_cols; 
    dim = X.n_rows;
    n_pool = n_trees * depth;
    n_array = pow(2, depth + 1);
    split_points = fmat();
    random_matrix = fmat();
}

/**
 * The function whose call starts the actual index construction. Initializes 
 * arrays to store the tree structures and computes all the projections needed
 * later. Then repeatedly calls method grow_subtree that builds a single 
 * RP-tree.
 */
void Mrpt::grow() {
    split_points = zeros<fmat>(n_array, n_trees);
    uvec indices = linspace<uvec>(0, n_samples - 1, n_samples);

    // generate the random matrix and project the data set onto it
    random_matrix = conv_to<fmat>::from(randn(n_pool, dim));
    projected_data = random_matrix * X;
    
    // Grow the trees
    for (int n_tree = 0; n_tree < n_trees; n_tree++) {       
        first_idx = n_tree * depth;
        std::vector<uvec> t = grow_subtree(indices, 0, 0, n_tree); 
        tree_leaves.push_back(t);
    }
}

/**
 * Builds a single random projection tree. The tree is constructed by recursively
 * projecting the data on a random vector and splitting into two by the median.
 * @param indices - The indices left in this branch
 * @param tree_level - The level in tree where the recursion is at
 * @param i - The index within the tree where we are at
 * @param n_tree - The index of the tree within the index
 * @return The leaves as a vector of arma::uvecs
 */
std::vector<uvec> Mrpt::grow_subtree(const uvec &indices, int tree_level, int i, uword n_tree) {
    int n = indices.size();
    int idx_left = 2 * i + 1;
    int idx_right = idx_left + 1;

    if (tree_level == depth) {
        std::vector<uvec> v;
        v.push_back(indices);
        return v;
    }

    uvec level = {first_idx + tree_level};
    frowvec projection = projected_data(level, indices);
    uvec ordered = sort_index(projection);

    int split_point = n % 2 ? n / 2 : n / 2 - 1; // median split
    int idx_split_point = ordered(split_point);
    int idx_split_point2 = ordered(split_point + 1);

    split_points(i, n_tree) = n % 2 ? projection(idx_split_point) : (projection(idx_split_point) + projection(idx_split_point2)) / 2;
    uvec left_indices = ordered.subvec(0, split_point);
    uvec right_indices = ordered.subvec(split_point + 1, n - 1);

    std::vector<uvec> v = grow_subtree(indices.elem(left_indices), tree_level + 1, idx_left, n_tree);
    std::vector<uvec> w = grow_subtree(indices.elem(right_indices), tree_level + 1, idx_right, n_tree);
    v.insert( v.end(), w.begin(), w.end() );
    return v;
}

/**
 * This function finds the k approximate nearest neighbors of the query object 
 * q. The accuracy of the query depends on both the parameters used for index 
 * construction and additional parameters given to this function. This 
 * function implements two tricks to improve performance. The voting trick 
 * interprets each index object in leaves returned by tree traversals as votes,
 * and only performs the final linear search with the 'elect' most voted 
 * objects. The priority queue trick keeps track of nodes where the split value
 * was close to the projection so that we can split the tree traversal to both
 * subtrees if we want.
 * @param q - The query object whose neighbors the function finds.
 * @param k - The number of neighbors the user wants the function to return
 * @param votes_required - The number of votes required for an object to be included in the linear search step
 * @param branches - The number of extra branches explored in the priority queue trick
 * @return The indices of the k approximate nearest neighbors in the original
 * data set for which the index was built.
 */
uvec Mrpt::query(const fvec& q, int k, int votes_required, int branches) {
    // Compute projections for all levels in all trees
    // initialize pq and vote-count vector
    fvec projected_query = random_matrix * q;
    uvec votes = zeros<uvec>(n_samples);
    std::priority_queue<Gap, std::vector<Gap>, std::greater<Gap>> pq;
    int j;
    
    // First push RP tree roots to priority queue with negative gap widths 
    // (will be handled before extra branches)
    for (int n_tree = 0; n_tree < n_trees; n_tree++){
        pq.push(Gap(n_tree, 0, n_tree*depth, -1));
    }
  
    // The tree traversal phase. At first the top of the queue will contain the 
    // 'n_trees' tree roots, after which 'b' extra branches will be explored
    for (int b = 0; b < branches + n_trees; b++){
        if (pq.empty()) break;
        Gap gap = pq.top();
        pq.pop();
        
        const fvec& tree = split_points.unsafe_col(gap.tree);
        int idx_tree = gap.node;
        int idx_left, idx_right;
        j = gap.level;
        double split_point = tree[idx_tree];

        while (split_point) {
            idx_left = 2 * idx_tree + 1;
            idx_right = idx_left + 1;
            if (projected_query(j) <= split_point) {
                idx_tree = idx_left;
                pq.push(Gap(gap.tree, idx_right, j+1, split_point-projected_query(j)));
            } else {
                idx_tree = idx_right;
                pq.push(Gap(gap.tree, idx_left, j+1, projected_query(j)-split_point));
            }
            j++;
            split_point = tree[idx_tree];
        }

        const uvec& idx_one_tree = tree_leaves[gap.tree][idx_tree - pow(2, depth) + 1];
        for (int idx : idx_one_tree)
            votes[idx]++;
    } 
    
    // The election phase. Choose the objects with enough votes.
    uvec elected(n_samples);
    do { 
        j = 0;
        for (int i=0; i<n_samples; i++){
            if (votes[i] >= votes_required){
                elected[j] = i;
                j++;
            }
        }
        // If not enough objects (at least k required) with enough votes 
        // decrease vote count requirement
        votes_required--;
    } while(j < k);
    elected.resize(j);
    
    // The linear search phase
    return exact_knn(X, q, k, elected);
}


/**
 * find k nearest neighbors from data for the query point
 * @param X - data matrix, row = data point, col = dimension
 * @param q - query point as a row matrix
 * @param k - number of neighbors searched for
 * @param indices - indices of the points in the original matrix where search is made
 * @return - indices of nearest neighbors in data matrix X as a column vector
 */
uvec exact_knn(const fmat& D, const fvec& q, uword k, uvec indices) {
    int n_cols = indices.size();
    fvec distances = fvec(n_cols);
    for (int i = 0; i < n_cols; i++)
        distances[i] = sum(pow((D.col(indices(i)) - q), 2));

    if(k == 1) {
        uvec ret(1);
        distances.min(ret[0]);
        return ret;
    }

    uvec sorted_indices = indices(sort_index(distances));
    return sorted_indices.size() > k ? sorted_indices.head(k): sorted_indices;
}